clang/lib/Parse/Parser.cpp

```diff
@@ -1049,6 +1049,7 @@ Parser::ParseExternalDeclaration(ParsedAttributes &Attrs,
 
   case tok::kw_extern:
     if (getLangOpts().CPlusPlus && NextToken().is(tok::kw_template)) {
+      ProhibitAttributes(Attrs);
```

Comment:
It is prohibited by the grammar for standard attributes, but it seems like it could be sensible to allow non-standard attributes?  WDYT?

Reply:
@erichkeane, thanks for the feedback. I followed this part of the spec

> No [attribute-specifier-seq](https://eel.is/c++draft/dcl.attr.grammar#nt:attribute-specifier-seq) ([[dcl.attr.grammar]](https://eel.is/c++draft/dcl.attr.grammar)) shall appertain to an [explicit instantiation](https://eel.is/c++draft/temp.explicit#3.sentence-3).

I also checked how GCC handles this case, and it seems that it disallows all kinds of attrs.

Should non-standard attributes be applied to this `S<int>` (_if we are talking about the current test case_) declaration?

Reply:
>I followed this part of the spec
Yep, understood.  But `attribute-specifier-seq` is ONLY the `[[...]]` spelling.  The REST is implementation defined extensions (that is, `__attribute__((..))` and `__declspec(..)` spellings).

That said, I spent some time looking, and both MSVC and GCC compilers seem to get thrown off pretty terribly by `<attribute-tokens> extern template`, so unless @AaronBallman can think of others, I think I'm ok with this.

I WOULD like to have our test check `__attribute__` and a `__declspec` variant as well though, so please add a test for those.

Reply:
ALSO: I checked and this should only prevent `regular` keywords-as-attributes, which seem to only be ARM based ones, so this won't affect things like __alignas  or calling conventions/etc.

Reply:
> I think it's reasonable to reject all attributes.
> 
> > ALSO: I checked and this should only prevent regular keywords-as-attributes, which seem to only be ARM based ones, so this won't affect things like __alignas or calling conventions/etc.
> 
> @erichkeane I don't follow this comment. I think we should reject all attributes, not keywords implemented as an attribute. Right?

AH, yes, sorry, I spent some time looking at this code and confirmed the above, it wasn't a request for the submitter.

This currently rejects all attributes and 'regular-keywords' (as spelled in `Attr.td`).  BUT those are only ARM ones, so those are fine IMO.

Reply:
Thanks! So yeah. we're agreed, the current approach is reasonable.